src/query/ts/m3db/encoded_block_iterator.go

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3db
+
+import (
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/functions/utils"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/storage"
+	"github.com/m3db/m3/src/query/ts/m3db/consolidators"
+)
+
+type consolidationSettings struct {
+	consolidationFn consolidators.ConsolidationFunc
+	currentTime     time.Time
+	bounds          models.Bounds
+}
+
+type encodedBlock struct {
+	// There is slightly different execution for the last block in the series
+	lastBlock            bool
+	meta                 block.Metadata
+	tagOptions           models.TagOptions
+	consolidation        consolidationSettings
+	seriesMetas          []block.SeriesMeta
+	seriesBlockIterators []encoding.SeriesIterator
+}
+
+// NewEncodedBlock builds an encoded block
+func NewEncodedBlock(
+	seriesBlockIterators []encoding.SeriesIterator,
+	bounds models.Bounds,
+	tagOptions models.TagOptions,
+	lastBlock bool,
+) (block.Block, error) {
+	consolidation := consolidationSettings{
+		consolidationFn: consolidators.TakeLast,
+		currentTime:     bounds.Start,
+		bounds:          bounds,
+	}
+
+	bl := newEncodedBlock(
+		seriesBlockIterators,
+		tagOptions,
+		consolidation,
+		lastBlock,
+	)
+	err := bl.generateMetas()
```

Comment:
```
if err := bl.generateMetas(); err != nil {
   return nil, err
}
```

Reply:
Don't know how valid it is, but a bit iffy about doing this pattern when the `err := XXX` has side effects